src/handlers/miner_hbbft_handler.erl

```diff
@@ -457,77 +484,189 @@ restore(OldState, NewState) ->
+    Threshold1 =
+        case Phase of
+            sig -> F + 1;
+            gossip -> Threshold0
+        end,
+    SigsValidRand =
+        lists:sublist(blockchain_utils:shuffle(SigsValidAll), Threshold1),
+    case
+        length(SigsValidAll) =< (3*F)+1 andalso
+        length(SigsValidRand) >= Threshold1
+    of
+        true ->
+            %% So, this is a little dicey, if we don't need all N signatures,
+            %% we might have competing subsets depending on message order.
+            %% Given that the underlying artifact they're signing is the same
+            %% though, it should be ok as long as we disregard the signatures
+            %% for testing equality but check them for validity
+            SigsValidRandSubset =
+                lists:sublist(lists:sort(SigsValidRand), Threshold0),
             case Phase of
                 gossip ->
-                    {ok, done, lists:sublist(lists:sort(ValidSignatures), Threshold0)};
+                    {{some, {done, SigsValidRandSubset}}, S0};
                 sig ->
-                    case length(ValidSignatures) >= Threshold0 of
+                    case length(SigsValidRand) >= Threshold0 of
                         true ->
-                            {ok, done, lists:sublist(lists:sort(ValidSignatures), Threshold0)};
+                            {{some, {done, SigsValidRandSubset}}, S0};
                         false ->
-                            {ok, gossip, lists:sort(ValidSignatures)}
+                            {{some, {gossip, lists:sort(SigsValidRand)}}, S0}
                     end
             end;
         false ->
-            false
+            {none, S0}
     end.
 
-filter_signatures(State=#state{artifact=Artifact, signatures=Signatures, members=Members}) ->
-    FilteredSignatures = lists:filter(fun({Address, Signature}) ->
-                         lists:member(Address, Members) andalso
-                         libp2p_crypto:verify(Artifact, Signature, libp2p_crypto:bin_to_pubkey(Address))
-                 end, Signatures),
-    State#state{signatures=FilteredSignatures}.
+-spec sig_is_valid(sig(), #state{}) -> boolean().
+sig_is_valid(
+    {<<Addr/binary>>, <<Sig/binary>>},
+    #state{artifact=Art, members=Mems, hbbft=HBBFT}
+) ->
+    IsVerified = libp2p_crypto:verify(Art, Sig, libp2p_crypto:bin_to_pubkey(Addr)),
+    IsMember = lists:member(Addr, Mems),
+    IsValid = IsMember andalso IsVerified,
+    case IsValid of
+        true ->
+            ok;
+        false ->
+            CurRound = hbbft:round(HBBFT),
+            % TODO Maybe this logging only makes sense in handle_msg_signature?
+            %lager:warning("Invalid signature ~s from ~s for round ~p in our round ~p", [b58(Sig), b58(Addr), SigRound, CurRound]),
+            lager:warning("Invalid signature ~s from ~s in our round ~p", [b58(Sig), b58(Addr), CurRound]),
+            lager:warning("member? ~p", [IsMember]),
+            lager:warning("verified? ~p", [IsVerified])
```

Comment:
@Vagabond If we remove this `verified?` logging - we can drop recomputing `libp2p_crypto:verify`.

Reply:
I'd prefer to infer it by looking at the membership check

Reply:
Without recomputing we can only go this far:
```erl
Verified =
    case IsMember of
        true -> member_with_invalid_sig;
        false -> non_member_with_unknown_sig_status
    end
```